mm2src/mm2_db/src/indexed_db/db_lock.rs

```diff
@@ -1,78 +1,123 @@
 pub type SharedDb<Db> = Arc<ConstructibleDb<Db>>;
 pub type WeakDb<Db> = Weak<ConstructibleDb<Db>>;
 
+#[allow(clippy::type_complexity)]
 pub struct ConstructibleDb<Db> {
     /// It's better to use something like [`Constructible`], but it doesn't provide a method to get the inner value by the mutable reference.
-    mutex: AsyncMutex<Option<Db>>,
+    locks: Arc<RwLock<HashMap<String, Arc<AsyncMutex<Option<Db>>>>>>,
     db_namespace: DbNamespaceId,
-    wallet_rmd160: Option<H160>,
+    // Default mm2 d_id derive from passphrase rmd160
+    db_id: String,
+    // Default mm2 shared_db_id derive from passphrase
+    shared_db_id: String,
 }
 
 impl<Db: DbInstance> ConstructibleDb<Db> {
     pub fn into_shared(self) -> SharedDb<Db> { Arc::new(self) }
 
     /// Creates a new uninitialized `Db` instance from other Iguana and/or HD accounts.
     /// This can be initialized later using [`ConstructibleDb::get_or_initialize`].
-    pub fn new(ctx: &MmArc) -> Self {
+    pub fn new(ctx: &MmArc, db_id: Option<&str>) -> Self {
+        let default_db_id = ctx.rmd160().to_string();
+        let shared_db_id = ctx.default_shared_db_id().to_string();
+
+        let db_id = db_id.unwrap_or(&default_db_id).to_string();
+        let conns = HashMap::from([(db_id.to_owned(), Arc::new(AsyncMutex::new(None)))]);
+
         ConstructibleDb {
-            mutex: AsyncMutex::new(None),
+            locks: Arc::new(RwLock::new(conns)),
             db_namespace: ctx.db_namespace,
-            wallet_rmd160: Some(*ctx.rmd160()),
+            db_id,
+            shared_db_id,
         }
     }
 
     /// Creates a new uninitialized `Db` instance shared between Iguana and all HD accounts
     /// derived from the same passphrase.
     /// This can be initialized later using [`ConstructibleDb::get_or_initialize`].
     pub fn new_shared_db(ctx: &MmArc) -> Self {
+        let db_id = hex::encode(ctx.rmd160().as_slice());
+        let shared_db_id = ctx.default_shared_db_id().to_string();
+        let conns = HashMap::from([(shared_db_id.clone(), Arc::new(AsyncMutex::new(None)))]);
         ConstructibleDb {
-            mutex: AsyncMutex::new(None),
+            locks: Arc::new(RwLock::new(conns)),
             db_namespace: ctx.db_namespace,
-            wallet_rmd160: Some(*ctx.shared_db_id()),
+            db_id,
+            shared_db_id,
         }
     }
 
     /// Creates a new uninitialized `Db` instance shared between all wallets/seed.
     /// This can be initialized later using [`ConstructibleDb::get_or_initialize`].
     pub fn new_global_db(ctx: &MmArc) -> Self {
+        let db_id = ctx.rmd160().to_string();
+        let shared_db_id = ctx.default_shared_db_id().to_string();
         ConstructibleDb {
-            mutex: AsyncMutex::new(None),
+            locks: Arc::new(RwLock::new(HashMap::default())),
             db_namespace: ctx.db_namespace,
-            wallet_rmd160: None,
+            db_id,
+            shared_db_id,
```

Comment:
Why the change here? `global_db` shouldn't have `db_id` or `shared_db_id`. This is used for wallets db which is a global db that stores encrypted mnemonics/seeds for all wallets. https://github.com/KomodoPlatform/komodo-defi-framework/blob/7a2ebafbd014a239bf1944a3fc8343b55373eb46/mm2src/mm2_main/src/lp_wallet.rs#L94
Related:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/7a2ebafbd014a239bf1944a3fc8343b55373eb46/mm2src/mm2_main/src/lp_wallet.rs#L99-L100

Reply:
fixed. thanks for the catch!

Reply:
I don't understand why the 2 below functions are needed and not only one
https://github.com/KomodoPlatform/komodo-defi-framework/blob/0856b54c4363f9247fc4e2b5dd304719ff9a2ac3/mm2src/mm2_db/src/indexed_db/indexed_db.rs#L113-L119

Reply:
seems like a leftover.. removed